XenAdmin/Wizards/RollingUpgradeWizard/PlanActions/RpuUploadAndApplySuppPackPlanAction.cs

```diff
@@ -0,0 +1,281 @@
+
+                    uploadIsoAction = new UploadSupplementalPackAction(connection, new List<Host> { master }, suppPackPath, true);
+                    uploadIsoAction.Changed += uploadAction_Changed;
+                    uploadIsoAction.Completed += uploadAction_Completed;
+                    uploadIsoAction.RunExternal(session);
+                }
+                catch (Exception ex)
+                {
+                    log.Error(string.Format("Upload failed for update {0} on connection {1}", suppPack, connection), ex);
+                    throw;
+                }
+
+                var poolupdate = uploadIsoAction.PoolUpdate;
+                if (poolupdate == null)
+                {
+                    log.ErrorFormat(
+                        "Upload finished successfully, but Pool_update object has not been found for update {0} on connection {1}.",
+                        suppPack, connection);
+                    throw new Exception(Messages.ACTION_UPLOADPATCHTOMASTERPLANACTION_FAILED);
+                }
+
+                uploadedSuppPacks.Add(master, poolupdate);
+            }
+            else if (host.uuid == master.uuid)
+            {
+                AddProgressStep(string.Format(Messages.UPDATES_WIZARD_SKIPPING_UPLOAD, suppPack, connection.Name));
+            }
+        }
+
+        private void PrecheckSuppPack(Pool_update update, string suppPack, out bool alreadyApplied)
+        {
+            alreadyApplied = false;
+            if (Cancelling)
+                throw new CancelledException();
+
+            var livePatchStatus = new Dictionary<string, livepatch_status>();
+
+            try
+            {
+                AddProgressStep(string.Format(Messages.UPDATES_WIZARD_RUNNING_PRECHECK, suppPack, host.Name()));
+
+
+                PatchPrecheckCheck check = new PatchPrecheckCheck(host, update, livePatchStatus);
+                var problems = check.RunAllChecks();
+                if (problems != null && problems.Count > 0)
+                {
+                    if (problems[0] is PatchAlreadyApplied)
+                    {
+                        log.InfoFormat("The update {0} is already applied on {1}. Ignore it.", suppPack, host.Name());
+                        ReplaceProgressStep(string.Format(Messages.UPDATES_WIZARD_SKIPPING_UPDATE, suppPack, host.Name()));
+                        alreadyApplied = true;
+                    }
+                    else
+                        throw new Exception(string.Format("{0}: {1}. {2}", host, problems[0].Title,
+                            problems[0].Description));
+                }
+            }
+            catch (Exception ex)
+            {
+                log.Error(string.Format("Precheck failed on host {0}", host.Name()), ex);
+                throw;
+            }
+
+            if (livePatchStatus[host.uuid] != livepatch_status.ok_livepatch_complete && !hostsThatWillRequireReboot.Contains(host.uuid))

```

Comment:
livePatchStatus[host.uuid] can throw a KeyNotFoundException, because a host is added to this dictionary in PatchPrecheckCheck only if the check succeeds. If it fails with an error (e.g. patch not applicable ar already applied), then it is not added.